File: tests/test_nx.c
/* test_nx.c - proves anonymous mmap() pages are really non-executable.
 *
 * Regression coverage for the NX/W^X hardening pass: sys_mmap() used
 * to ignore `prot` entirely and map every anonymous page executable.
 * This mmaps a page (no PROT_EXEC), writes a `ret` (0xc3) into it,
 * and jumps to it. On a fixed kernel this must fault and kill the
 * task - if it prints "UNSAFE" instead, NX enforcement regressed.
 */
#include <sys/mman.h>
#include <stdio.h>
#include <string.h>

typedef void (*fn)(void);

int main(void) {
    void *p = mmap(NULL, 4096, PROT_READ | PROT_WRITE,
                   MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
    if (p == MAP_FAILED) {
        printf("nxtest: mmap failed\n");
        return 1;
    }
    ((unsigned char *)p)[0] = 0xc3; /* ret */
    printf("nxtest: about to jump to a non-executable page\n");
    ((fn)p)();
    /* Only reached if the NX bit was NOT enforced. */
    printf("nxtest: UNSAFE - executed a non-executable page\n");
    return 0;
}

File: tests/test_nx.py
#!/usr/bin/env python3
"""
test_nx.py - proves NX/W^X enforcement is real, not just present in
the source.

Regression coverage for the NX hardening pass (EFER.NXE + VMM_NX on
ELF data segments, user stacks, and anonymous mmap() without
PROT_EXEC): `nxtest` (tests/test_nx.c) mmaps an anonymous page,
writes a `ret` opcode into it, and jumps to it. On a correctly
hardened kernel the CPU raises a page fault (NX violation) the
instant execution reaches that page, and TUS kills the task instead
of running attacker-controlled bytes as code. If nxtest instead
prints "UNSAFE", the mmap NX default regressed.
"""

import sys

sys.path.insert(0, "tests")
import test_boot  # noqa: E402
from test_boot import start_qemu, qmp_connect, type_text, wait_for  # noqa: E402


def main():
    qemu = start_qemu()
    try:
        sock = qmp_connect()

        offset = wait_for("tsh ready", timeout=60)
        offset = wait_for("graphics test be performed", offset=offset,
                          timeout=15)
        type_text(sock, "n\r")
        try:
            wait_for("login:", timeout=15, offset=offset)
            type_text(sock, "root\r")
            wait_for("Password:", offset=offset, timeout=15)
            type_text(sock, "toast\r")
        except AssertionError:
            pass
        offset = wait_for("tus:/>", offset=offset)

        type_text(sock, "nxtest\r")
        offset = wait_for("about to jump to a non-executable page",
                          offset=offset, timeout=15)
        offset = wait_for("killed:", offset=offset, timeout=15)
        print("PASS: an anonymous mmap() page without PROT_EXEC is "
              "genuinely non-executable - jumping to it faults and "
              "kills the task")
        return 0
    finally:
        qemu.terminate()
        qemu.wait(timeout=10)


if __name__ == "__main__":
    sys.exit(main())
